File: packages/physics-core/src/cathsim_physics/frames.py
"""Adapted (Bishop) frame utilities for the discrete elastic rod.

A Bishop frame is a rotation-minimising orthonormal frame ``(t, d1, d2)``
carried along the rod centreline.  It is obtained by parallel-transporting the
director ``d1`` of the first edge onto every following edge.  By construction a
Bishop frame carries **zero twist**, which is exactly why we use it: the twist
degree of freedom ``phi`` of an edge is then the *material* twist angle
measured against a twist-free reference, and the twist energy becomes a
function of ``phi`` alone (see :mod:`cathsim_physics.rod`).

All vectors are in millimetres / dimensionless directions.

Research prototype - Not for clinical use.
"""

from __future__ import annotations

import numpy as np

_EPS = 1.0e-12


def normalise(vectors: np.ndarray) -> np.ndarray:
    """Return ``vectors`` (..., 3) scaled to unit length.

    Raises:
        ValueError: if any vector is (numerically) of zero length, which in a
            rod means a collapsed edge and is never physically meaningful.
    """
    norms = np.linalg.norm(vectors, axis=-1, keepdims=True)
    if not np.all(np.isfinite(norms)) or np.any(norms < _EPS):
        raise ValueError(
            "Cannot normalise a zero-length or non-finite direction vector. "
            "This usually means two rod nodes collapsed onto each other; "
            "reduce the time step or increase the axial stiffness."
        )
    return vectors / norms


def parallel_transport(vector: np.ndarray, t_from: np.ndarray, t_to: np.ndarray) -> np.ndarray:
    """Parallel-transport ``vector`` from unit tangent ``t_from`` to ``t_to``.

    Rotation-minimising transport about the axis ``t_from x t_to``.  Works on a
    single 3-vector or on a batch of shape ``(..., 3)``.
    """
    vector = np.asarray(vector, dtype=float)
    t_from = np.asarray(t_from, dtype=float)
    t_to = np.asarray(t_to, dtype=float)

    axis = np.cross(t_from, t_to)
    sin_theta = np.linalg.norm(axis, axis=-1, keepdims=True)
    cos_theta = np.sum(t_from * t_to, axis=-1, keepdims=True)

    # Degenerate case: tangents (anti)parallel -> transport is the identity.
    safe = sin_theta > _EPS
    axis_unit = np.where(safe, axis / np.where(safe, sin_theta, 1.0), 0.0)

    # Rodrigues rotation.
    cross_term = np.cross(axis_unit, vector)
    dot_term = np.sum(axis_unit * vector, axis=-1, keepdims=True)
    rotated = vector * cos_theta + cross_term * sin_theta + axis_unit * dot_term * (1.0 - cos_theta)
    return np.where(safe, rotated, vector)


def any_perpendicular(tangent: np.ndarray) -> np.ndarray:
    """Return an arbitrary unit vector perpendicular to ``tangent`` (3,)."""
    tangent = normalise(np.asarray(tangent, dtype=float))
    helper = np.array([0.0, 0.0, 1.0])
    if abs(float(np.dot(tangent, helper))) > 0.9:
        helper = np.array([1.0, 0.0, 0.0])
    return normalise(np.cross(tangent, helper))


def bishop_frame(tangents: np.ndarray, d1_first: np.ndarray) -> np.ndarray:
    """Build the Bishop directors ``d1`` for every edge.

    Args:
        tangents: ``(M, 3)`` unit tangents, one per edge.
        d1_first: ``(3,)`` director of edge 0, must be perpendicular to
            ``tangents[0]`` (it is re-orthogonalised defensively).

    Returns:
        ``(M, 3)`` array of unit directors, each perpendicular to its tangent.
    """
    tangents = np.asarray(tangents, dtype=float)
    n_edges = tangents.shape[0]
    d1 = np.empty((n_edges, 3), dtype=float)

    first = np.asarray(d1_first, dtype=float)
    first = first - np.dot(first, tangents[0]) * tangents[0]
    d1[0] = normalise(first)

    for edge in range(1, n_edges):
        transported = parallel_transport(d1[edge - 1], tangents[edge - 1], tangents[edge])
        transported = transported - np.dot(transported, tangents[edge]) * tangents[edge]
        d1[edge] = normalise(transported)
    return d1


def material_frame(
    tangents: np.ndarray, d1: np.ndarray, phi_rad: np.ndarray
) -> tuple[np.ndarray, np.ndarray]:
    """Rotate the Bishop directors by the material twist angle ``phi``.

    Returns ``(m1, m2)``, each ``(M, 3)``::

        m1 =  cos(phi) d1 + sin(phi) d2
        m2 = -sin(phi) d1 + cos(phi) d2      with d2 = t x d1
    """
    d2 = np.cross(tangents, d1)
    cos_phi = np.cos(phi_rad)[:, None]
    sin_phi = np.sin(phi_rad)[:, None]
    m1 = cos_phi * d1 + sin_phi * d2
    m2 = -sin_phi * d1 + cos_phi * d2
    return m1, m2


def frame_to_quaternion_xyzw(tangent: np.ndarray, m1: np.ndarray, m2: np.ndarray) -> np.ndarray:
    """Convert the orthonormal frame ``(m1, m2, t)`` to quaternions ``[x,y,z,w]``.

    Column convention: the rotation matrix maps the local axes
    ``(e_x, e_y, e_z)`` onto ``(m1, m2, tangent)``.  Batched over ``(M, 3)``.
    """
    m = np.stack([m1, m2, tangent], axis=-1)  # (M, 3, 3), columns are the axes
    trace = m[:, 0, 0] + m[:, 1, 1] + m[:, 2, 2]
    quat = np.zeros((m.shape[0], 4), dtype=float)

    positive = trace > 0.0
    s = np.zeros(m.shape[0])
    s[positive] = np.sqrt(trace[positive] + 1.0) * 2.0
    with np.errstate(invalid="ignore", divide="ignore"):
        quat[positive, 3] = 0.25 * s[positive]
        quat[positive, 0] = (m[positive, 2, 1] - m[positive, 1, 2]) / s[positive]
        quat[positive, 1] = (m[positive, 0, 2] - m[positive, 2, 0]) / s[positive]
        quat[positive, 2] = (m[positive, 1, 0] - m[positive, 0, 1]) / s[positive]

    # Fallback branch for trace <= 0: pick the largest diagonal element.
    for index in np.flatnonzero(~positive):
        mat = m[index]
        diag = np.argmax([mat[0, 0], mat[1, 1], mat[2, 2]])
        if diag == 0:
            root = np.sqrt(1.0 + mat[0, 0] - mat[1, 1] - mat[2, 2]) * 2.0
            quat[index] = [
                0.25 * root,
                (mat[0, 1] + mat[1, 0]) / root,
                (mat[0, 2] + mat[2, 0]) / root,
                (mat[2, 1] - mat[1, 2]) / root,
            ]
        elif diag == 1:
            root = np.sqrt(1.0 + mat[1, 1] - mat[0, 0] - mat[2, 2]) * 2.0
            quat[index] = [
                (mat[0, 1] + mat[1, 0]) / root,
                0.25 * root,
                (mat[1, 2] + mat[2, 1]) / root,
                (mat[0, 2] - mat[2, 0]) / root,
            ]
        else:
            root = np.sqrt(1.0 + mat[2, 2] - mat[0, 0] - mat[1, 1]) * 2.0
            quat[index] = [
                (mat[0, 2] + mat[2, 0]) / root,
                (mat[1, 2] + mat[2, 1]) / root,
                0.25 * root,
                (mat[1, 0] - mat[0, 1]) / root,
            ]
    return quat / np.linalg.norm(quat, axis=-1, keepdims=True)

File: apps/viewer/src/physics/frames.ts
/**
 * Bishop (rotation-minimising) frame utilities - TypeScript port of
 * `packages/physics-core/src/cathsim_physics/frames.py`.
 *
 * A Bishop frame carries zero twist by construction, so the per-edge angle
 * `phi` is the material twist measured against a twist-free reference.
 *
 * Research prototype - Not for clinical use.
 */

import { nodeCount, type RodState } from './rod';

const EPSILON = 1e-12;

export interface RodFrames {
  /** Unit tangents, flat 3(N-1). */
  tangents: Float64Array;
  /** Bishop directors, flat 3(N-1). */
  d1: Float64Array;
  /** Material director m1 = cos(phi) d1 + sin(phi) d2, flat 3(N-1). */
  m1: Float64Array;
  /** Material director m2 = -sin(phi) d1 + cos(phi) d2, flat 3(N-1). */
  m2: Float64Array;
}

/** Rotate `v` from unit tangent `from` onto unit tangent `to` (Rodrigues). */
export function parallelTransport(
  v: readonly number[],
  from: readonly number[],
  to: readonly number[],
): number[] {
  const ax = from[1] * to[2] - from[2] * to[1];
  const ay = from[2] * to[0] - from[0] * to[2];
  const az = from[0] * to[1] - from[1] * to[0];
  const sin = Math.hypot(ax, ay, az);
  if (sin < EPSILON) return [v[0], v[1], v[2]];
  const cos = from[0] * to[0] + from[1] * to[1] + from[2] * to[2];
  const ux = ax / sin;
  const uy = ay / sin;
  const uz = az / sin;
  const cx = uy * v[2] - uz * v[1];
  const cy = uz * v[0] - ux * v[2];
  const cz = ux * v[1] - uy * v[0];
  const d = (ux * v[0] + uy * v[1] + uz * v[2]) * (1 - cos);
  return [v[0] * cos + cx * sin + ux * d, v[1] * cos + cy * sin + uy * d, v[2] * cos + cz * sin + uz * d];
}

/** Compute tangents, Bishop directors and the material frame of a rod state. */
export function computeFrames(state: RodState): RodFrames {
  const n = nodeCount(state);
  const edges = n - 1;
  const tangents = new Float64Array(3 * edges);
  const d1 = new Float64Array(3 * edges);
  const m1 = new Float64Array(3 * edges);
  const m2 = new Float64Array(3 * edges);
  const x = state.nodesMm;

  for (let j = 0; j < edges; j += 1) {
    const a = 3 * j;
    const b = 3 * (j + 1);
    const ex = x[b] - x[a];
    const ey = x[b + 1] - x[a + 1];
    const ez = x[b + 2] - x[a + 2];
    const len = Math.hypot(ex, ey, ez) || 1;
    tangents[a] = ex / len;
    tangents[a + 1] = ey / len;
    tangents[a + 2] = ez / len;
  }

  // Anchor: orthogonalise the requested director against the first tangent.
  let previous = orthonormalise(
    [state.d1First[0], state.d1First[1], state.d1First[2]],
    [tangents[0], tangents[1], tangents[2]],
  );
  d1.set(previous, 0);
  for (let j = 1; j < edges; j += 1) {
    const from = [tangents[3 * (j - 1)], tangents[3 * (j - 1) + 1], tangents[3 * (j - 1) + 2]];
    const to = [tangents[3 * j], tangents[3 * j + 1], tangents[3 * j + 2]];
    previous = orthonormalise(parallelTransport(previous, from, to), to);
    d1.set(previous, 3 * j);
  }

  for (let j = 0; j < edges; j += 1) {
    const a = 3 * j;
    const tx = tangents[a];
    const ty = tangents[a + 1];
    const tz = tangents[a + 2];
    const d1x = d1[a];
    const d1y = d1[a + 1];
    const d1z = d1[a + 2];
    // d2 = t x d1
    const d2x = ty * d1z - tz * d1y;
    const d2y = tz * d1x - tx * d1z;
    const d2z = tx * d1y - ty * d1x;
    const cos = Math.cos(state.phiRad[j]);
    const sin = Math.sin(state.phiRad[j]);
    m1[a] = cos * d1x + sin * d2x;
    m1[a + 1] = cos * d1y + sin * d2y;
    m1[a + 2] = cos * d1z + sin * d2z;
    m2[a] = -sin * d1x + cos * d2x;
    m2[a + 1] = -sin * d1y + cos * d2y;
    m2[a + 2] = -sin * d1z + cos * d2z;
  }

  return { tangents, d1, m1, m2 };
}

function orthonormalise(v: readonly number[], axis: readonly number[]): number[] {
  const projection = v[0] * axis[0] + v[1] * axis[1] + v[2] * axis[2];
  let px = v[0] - projection * axis[0];
  let py = v[1] - projection * axis[1];
  let pz = v[2] - projection * axis[2];
  let norm = Math.hypot(px, py, pz);
  if (norm < EPSILON) {
    // Degenerate: pick any perpendicular.
    const helper = Math.abs(axis[2]) > 0.9 ? [1, 0, 0] : [0, 0, 1];
    px = axis[1] * helper[2] - axis[2] * helper[1];
    py = axis[2] * helper[0] - axis[0] * helper[2];
    pz = axis[0] * helper[1] - axis[1] * helper[0];
    norm = Math.hypot(px, py, pz);
  }
  return [px / norm, py / norm, pz / norm];
}

/** Quaternion [x,y,z,w] of the frame whose columns are (m1, m2, tangent). */
export function frameToQuaternion(
  m1: readonly number[],
  m2: readonly number[],
  tangent: readonly number[],
): [number, number, number, number] {
  const m00 = m1[0];
  const m10 = m1[1];
  const m20 = m1[2];
  const m01 = m2[0];
  const m11 = m2[1];
  const m21 = m2[2];
  const m02 = tangent[0];
  const m12 = tangent[1];
  const m22 = tangent[2];
  const trace = m00 + m11 + m22;
  let q: [number, number, number, number];
  if (trace > 0) {
    const s = Math.sqrt(trace + 1) * 2;
    q = [(m21 - m12) / s, (m02 - m20) / s, (m10 - m01) / s, 0.25 * s];
  } else if (m00 > m11 && m00 > m22) {
    const s = Math.sqrt(1 + m00 - m11 - m22) * 2;
    q = [0.25 * s, (m01 + m10) / s, (m02 + m20) / s, (m21 - m12) / s];
  } else if (m11 > m22) {
    const s = Math.sqrt(1 + m11 - m00 - m22) * 2;
    q = [(m01 + m10) / s, 0.25 * s, (m12 + m21) / s, (m02 - m20) / s];
  } else {
    const s = Math.sqrt(1 + m22 - m00 - m11) * 2;
    q = [(m02 + m20) / s, (m12 + m21) / s, 0.25 * s, (m10 - m01) / s];
  }
  const norm = Math.hypot(q[0], q[1], q[2], q[3]);
  return [q[0] / norm, q[1] / norm, q[2] / norm, q[3] / norm];
}
